walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,181 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Sort Collection by start time by using a stream
+    events.stream().sorted((e1, e2) -> e1.getWhen().start() - e2.getWhen().start());
+
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Collection<String> mandatoryAttendees = request.getAttendees();
+
+    // No meeting attendees: whole day is open
+    if (optionalAttendees.isEmpty() && mandatoryAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.fromStartEnd(START_OF_DAY, END_OF_DAY, false));
+    }
+
+    // First track times unavailable for the group
+    List<TimeRange> unavailableTimes = new ArrayList<>();
+    List<TimeRange> optionalUnavailableTimes = new ArrayList<>();
+    for (Event e : events) {
+      // If the attendee lists are not disjoint (meaning there is at least one attendee in common), 
+      // save this event time as unavailable
+      if (!Collections.disjoint(e.getAttendees(), mandatoryAttendees)) {
+        unavailableTimes.add(e.getWhen());
+      } 
+      if (!Collections.disjoint(e.getAttendees(), optionalAttendees)) {
+        optionalUnavailableTimes.add(e.getWhen());
+      } 
+    }
+
+    // Coalesce any overlapping times in the lists
+    coalesceList(unavailableTimes);
+    coalesceList(optionalUnavailableTimes);
+
+    List<TimeRange> availableTimes = new ArrayList<>();
+    List<TimeRange> optionalAvailableTimes = new ArrayList<>();
+    long duration = request.getDuration();
+
+    if (!unavailableTimes.isEmpty()) {
+      // Check if any times are open for mandatory attendees
+      // Methods add open times to availableTimes list
+      checkMorning(unavailableTimes, availableTimes, duration);
+      checkBetweenMeetings(unavailableTimes, availableTimes, duration);
+      checkEvening(unavailableTimes, availableTimes, duration);
+    }
+    else if (!mandatoryAttendees.isEmpty() && duration <= END_OF_DAY) {
+      // Entire day is open
+      availableTimes.add(TimeRange.fromStartEnd(START_OF_DAY, END_OF_DAY, false));
+    }
+
+    if (!optionalUnavailableTimes.isEmpty()) {
+      // Check if any times are open for optional attendees
+      // Methods add open times to optionalAvailableTimes list
+      checkMorning(optionalUnavailableTimes, optionalAvailableTimes, duration);
+      checkBetweenMeetings(optionalUnavailableTimes, optionalAvailableTimes, duration);
+      checkEvening(optionalUnavailableTimes, optionalAvailableTimes, duration);
+    }
+    else if (!optionalAttendees.isEmpty() && duration <= END_OF_DAY) {
+      // Entire day is open
+      optionalAvailableTimes.add(TimeRange.fromStartEnd(START_OF_DAY, END_OF_DAY, false));
+    }
+
+    // Attempt to combine to two lists to see if there are times when mandatory
+    // and optional attendees are both available
+    List<TimeRange> totalAvailableTimes = combineAvailableLists(availableTimes, optionalAvailableTimes, duration);
+
```

Comment:
I think you might be complicating your life a bit in your implementation of this method. I had a hard time following the if block below, because of the interplay of all the artifacts you generate above.

Consider trying some logic that looks more like:

```
// Just pseudo-code to show the logic flow... and you'd have to implement invert.
// invert is just "all time windows NOT in the sorted event lists" - note that you don't
// need to coalesce if they're sorted... just follow the chain until there's a gap between two events.
allAvailableTimes = filterTimeWindowLength(invert(allEvents));
allRequiredAvailableTimes = filterTimeWindowLength(invert(allRequiredEvents));
return allAvailableTimes.isEmpty()
    ? allRequiredAvailableTimes
    : allAvailableTimes;
```
(unless you want to get into fancy sorting by "number of people who can't make it" or somesuch later)

Reply:
This implementation makes so much more sense; thank you! I've implemented it in the latest commit.
I still ended up having to coalesce the lists of times where there were events, because if I had an all-day event for someone with two smaller events contained within it, for example, the invert method would end up adding one or more of the gaps between the smaller events as I was following the chain of the 3+ events. It worked once I coalesced everything together, but please let me know if there was something I could have done differently in invert that would have eliminated the need for coalesce, because I could not think of a way.
Thank you again for your comments and helping me really clean up the code!